caravan.lua

```diff
@@ -27,6 +27,312 @@ specified, then the commands apply to all caravans on the map.
 
 --@ module = true
 
+local gui = require('gui')
+local overlay = require('plugins.overlay')
+local widgets = require('gui.widgets')
+
+trader_selected_state = trader_selected_state or {}
+broker_selected_state = broker_selected_state or {}
+handle_ctrl_click_on_render = handle_ctrl_click_on_render or false
+handle_shift_click_on_render = handle_shift_click_on_render or false
+
+dfhack.onStateChange.caravanOverlay = function(code)
+    if code == SC_WORLD_UNLOADED then
+        trader_selected_state = {}
+        broker_selected_state = {}
+        handle_ctrl_click_on_render = false
+        handle_shift_click_on_render = false
+    end
+end
+
+local GOODFLAG = {
+    UNCONTAINED_UNSELECTED = 0,
+    UNCONTAINED_SELECTED = 1,
+    CONTAINED_UNSELECTED = 2,
+    CONTAINED_SELECTED = 3,
+    CONTAINER_COLLAPSED_UNSELECTED = 4,
+    CONTAINER_COLLAPSED_SELECTED = 5,
+}
+
+function is_trade_overlay_active()
+    return df.global.game.main_interface.trade.open
+end
+
+CaravanOverlay = defclass(CaravanOverlay, overlay.OverlayWidget)
```

Comment:
every overlay we define can be individually enabled/disabled by the player. Is there any reason this is separate from the CaravanInfoOverlay below? As it is, it looks like they should be treated as a single unit.

Reply:
Yeah, they should be a single unit for sure. Just inexperience with overlay stuff there. I'll fix it up.